ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -178,102 +183,208 @@ newtype Our a = Our { unOur :: a }
+  NoThunks (Header blk)
+  ) => NoThunks (ChainSyncStateView m blk)
 bracketChainSyncClient ::
     ( IOLike m
     , Ord peer
     , LedgerSupportsProtocol blk
     )
  => Tracer m (TraceChainSyncClientEvent blk)
  -> ChainDbView m blk
- -> StrictTVar m (Map peer (StrictTVar m (AnchoredFragment (Header blk))))
-    -- ^ The candidate chains, we need the whole map because we
-    -- (de)register nodes (@peer@).
- -> StrictTVar m (Set peer)
-    -- ^ This ChainSync client should ensure that its peer is in this set while
-    -- and only while both of the following conditions are satisfied: the
-    -- peer's latest -- message has been fully processed (especially that its
-    -- candidate has been updated; previous argument) and its latest message
-    -- did not claim that it already has headers that extend its candidate.
-    --
-    -- It's more important that the client is removed from the set promptly
-    -- than it is for the client to be added promptly, because of how this is
-    -- used by the GSM to determine that the node is done syncing.
  -> StrictTVar m (Map peer (ChainSyncClientHandle m blk))
+    -- ^ The kill handle and states for each peer, we need the whole map because we
+    -- (de)register nodes (@peer@).
  -> peer
  -> NodeToNodeVersion
  -> ChainSyncLoPBucketConfig
- -> (     StrictTVar m (AnchoredFragment (Header blk))
-          -- ^ Variable holding the current fragment
-       -> (m (), m ())
-       -> (m (), m (), m ())
-       -> (Their (Tip blk) -> STM m ())
-          -- ^ callback to set the last announced tip
-       -> (SlotNo -> STM m ())
-          -- ^ callback to set the slot of the last received header
-       -> m a
-    )
+ -> (ChainSyncStateView m blk -> m a)
  -> m a
 bracketChainSyncClient
     tracer
     ChainDbView { getIsInvalidBlock }
-    varCandidates
-    varIdling
     varHandles
     peer
     version
     csBucketConfig
     body
   =
-    bracket newCandidateVar releaseCandidateVar
-  $ \(varCandidate, setTheirTip, setLatestSlot) ->
+    bracket acquireHandle releaseHandle
+  $ \varState ->
         withWatcher
             "ChainSync.Client.rejectInvalidBlocks"
-            (invalidBlockWatcher varCandidate)
+            (invalidBlockWatcher varState)
       $ LeakyBucket.execAgainstBucket lopBucketConfig
       $ \lopBucket ->
-            body
-                varCandidate
-                ( atomically $ modifyTVar varIdling $ Set.insert peer
-                , atomically $ modifyTVar varIdling $ Set.delete peer )
-                ( LeakyBucket.setPaused lopBucket True
-                , LeakyBucket.setPaused lopBucket False
-                , void $ LeakyBucket.fill lopBucket 1 )
-                setTheirTip
-                setLatestSlot
+            body ChainSyncStateView {
+              csvSetCandidate =
+              modifyTVar varState . \ c s -> s {csCandidate = c}
+            , csvSetLatestSlot =
+              modifyTVar varState . \ ls s -> s {csLatestSlot = Just ls}
```

Comment:
```suggestion
              modifyTVar varState . \ ls s -> s {csLatestSlot = Just $! ls}
```
I *think* it shouldn't matter given how we invoke `csvSetLatestSlot`, but this should make sure that `ls` is forced to WHNF such that the nothunks tests don't complain. Alternative would be to use `StrictMaybe` from `cardano-strict-containers`.

Reply:
97619e6ee885e5f038dabb3c1827d8205491d084